### PR TITLE
[ed25519] Use `verify_strict` for signature verification in ed25519 precompile

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
[simd 152](https://github.com/solana-foundation/solana-improvement-documents/pull/152) is merged. This PR is change # 1, which updates ed25519 precompile signature verification to use the strict verification.

#### Summary of Changes
Replaced signature verification in the ed25519 precompile to use strict verification under a feature gate.

Fixes #

### CODE REVIEW

**Review by seanyoung — COMMENTED** [review]
Looks good, but I think it needs a test case.

**Comment by 0x0ece** [comment]
Example test:
https://github.com/0x0ece/agave/commit/ba98faacb61ce19862ee2edfe277fdaf9c10432d

FYI, in solfuzz-agave & [test-vectors](https://github.com/firedancer-io/test-vectors/tree/main/precompile/fixtures/ed25519) we have already 1k+ tests (all wycheproof + all cctv), so no need to be too detailed here. Just make sure the feat gate works.

**Review by seanyoung — DISMISSED** [review]
Looks great

**Comment by t-nelson** [comment]
code looks good. we should probably remeasure runtime to be sure the CU price doesn't need adjusted

**Comment by samkim-crypto** [comment]
I ran the [benches](https://github.com/anza-xyz/agave/blob/master/sdk/benches/ed25519_instructions.rs) on my devserver and it does seem that `verify_strict` increases cost by ~4-5%.

Standard verification
```
test bench_ed25519_len_032 ... bench:      53,174.08 ns/iter (+/- 320.90)
test bench_ed25519_len_128 ... bench:      53,227.03 ns/iter (+/- 460.80)
test bench_ed25519_len_32k ... bench:     105,673.43 ns/iter (+/- 215.66)
test bench_ed25519_len_max ... bench:     157,801.53 ns/iter (+/- 393.42)
```

Strict verification
```
test bench_ed25519_len_032 ... bench:      56,098.69 ns/iter (+/- 652.68)
test bench_ed25519_len_128 ... bench:      56,373.39 ns/iter (+/- 408.99)
test bench_ed25519_len_32k ... bench:     109,504.36 ns/iter (+/- 2,027.92)
test bench_ed25519_len_max ... bench:     163,012.00 ns/iter (+/- 644.40)
```

**Comment by 0x0ece** [comment]
The bench is expected, however all precompiles cost like a tx sigverify, and sigverify is already using verify_strict.

If anything, users were charged 4-5% extra before, after this change they’ll be charged “correctly” (And are prob undercharged when using secp256k1, but this is off topic here)

**Review by seanyoung — DISMISSED** [review]

**Comment by samkim-crypto** [comment]
Hm, it does seem like the CU price for sigverify and precompile ed25519 are assigned different values in the cost model. @tao-stones, can you help if the current cost model for ed25519 is adequate? Looking at the commit history of the cost model, I can't seem to find how the cost models were originally assigned.

**Comment by 0x0ece** [comment]
Do you have ref to the code? TBH I was quoting something I saw in [firedancer](https://github.com/firedancer-io/firedancer/blob/main/src/flamenco/runtime/fd_runtime.c#L2654), so maybe we have implemented the wrong thing...

**Comment by tao-stones** [comment]
> Hm, it does seem like the CU price for sigverify and precompile ed25519 are assigned different values in the cost model. @tao-stones, can you help if the current cost model for ed25519 is adequate? Looking at the commit history of the cost model, I can't seem to find how the cost models were originally assigned.

Hey @samkim-crypto, precompiled costs were determined by benching (under `sdk/benches`). Transaction sigverify process is optimized, so it'd have different performance than ed25519 or secp256k1, iirc.

**Comment by samkim-crypto** [comment]
> Do you have ref to the code? TBH I was quoting something I saw in [firedancer](https://github.com/firedancer-io/firedancer/blob/main/src/flamenco/runtime/fd_runtime.c#L2654), so maybe we have implemented the wrong thing...

@0x0ece, I believe the costs for the precompiles are defined in the `solana-cost-model` crate [(here](https://github.com/anza-xyz/agave/blob/master/cost-model/src/block_cost_limits.rs#L30))

> > Hm, it does seem like the CU price for sigverify and precompile ed25519 are assigned different values in the cost model. @tao-stones, can you help if the current cost model for ed25519 is adequate? Looking at the commit history of the cost model, I can't seem to find how the cost models were originally assigned.
> 
> Hey @samkim-crypto, precompiled costs were determined by benching (under `sdk/benches`). Transaction sigverify process is optimized, so it'd have different performance than ed25519 or secp256k1, iirc.

@tao-stones, if I run the [benches](https://github.com/anza-xyz/agave/blob/master/sdk/benches/ed25519_instructions.rs), I consistency get about 4-5% increase in the timing both in my local and dev machines. Do you think it is reasonable to increase the constant factor [here](https://github.com/anza-xyz/agave/blob/master/cost-model/src/block_cost_limits.rs#L30) by 5%?

**Comment by tao-stones** [comment]
> if I run the [benches](https://github.com/anza-xyz/agave/blob/master/sdk/benches/ed25519_instructions.rs), I consistency get about 4-5% increase in the timing both in my local and dev machines. Do you think it is reasonable to increase the constant factor [here](https://github.com/anza-xyz/agave/blob/master/cost-model/src/block_cost_limits.rs#L30) by 5%?

the message length matters, currently tx can have 1K at most. iirc, it was benched with 3k message len to be reasonably conservative. Assuming 5% increase you are seeing are with different message len?

**Comment by samkim-crypto** [comment]
Yeah I did some more formal testing and I seem to be getting about 4% max increase on any message lengths on my dev machine. It does seem reasonable to increase the cost [constant](https://github.com/anza-xyz/agave/blob/master/cost-model/src/block_cost_limits.rs#L30) by 4%, so either 79 (maybe 80).

Standard verification
```
test bench_ed25519_len_032 ... bench:      53,239.98 ns/iter (+/- 498.79)
test bench_ed25519_len_128 ... bench:      53,477.30 ns/iter (+/- 370.17)
test bench_ed25519_len_1k  ... bench:      54,969.69 ns/iter (+/- 329.04)
test bench_ed25519_len_3k  ... bench:      58,245.12 ns/iter (+/- 245.52) <-- 3k
test bench_ed25519_len_5k  ... bench:      61,531.41 ns/iter (+/- 260.65)
test bench_ed25519_len_8k  ... bench:      66,428.60 ns/iter (+/- 229.59)
test bench_ed25519_len_16k ... bench:      79,600.47 ns/iter (+/- 335.06)
test bench_ed25519_len_32k ... bench:     105,824.82 ns/iter (+/- 339.31)
test bench_ed25519_len_max ... bench:     157,929.05 ns/iter (+/- 576.27)
```

Strict verification
```
test bench_ed25519_len_032 ... bench:      54,971.81 ns/iter (+/- 972.73) <-- 3.25% increase
test bench_ed25519_len_128 ... bench:      55,226.82 ns/iter (+/- 730.51) <-- 3,27% increase
test bench_ed25519_len_1k  ... bench:      56,739.93 ns/iter (+/- 1,401.09) <-- 3.22% increase
test bench_ed25519_len_3k  ... bench:      59,967.48 ns/iter (+/- 284.36) <-- 2.95% increase
test bench_ed25519_len_5k  ... bench:      63,248.71 ns/iter (+/- 211.82) <-- 3.79% increase
test bench_ed25519_len_8k  ... bench:      68,155.69 ns/iter (+/- 330.51) <-- 2.60% increase
test bench_ed25519_len_16k ... bench:      81,229.28 ns/iter (+/- 283.99) <-- 2.05% increase
test bench_ed25519_len_32k ... bench:     107,439.18 ns/iter (+/- 405.38) <-- 1.53% increase
test bench_ed25519_len_max ... bench:     159,484.30 ns/iter (+/- 344.30) <-- 0.98% increase
```

File: cost-model/src/block_cost_limits.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ pub const SIGNATURE_COST: u64 = COMPUTE_UNIT_TO_US_RATIO * 24;
 /// Number of compute units for one secp256k1 signature verification.
 pub const SECP256K1_VERIFY_COST: u64 = COMPUTE_UNIT_TO_US_RATIO * 223;
 /// Number of compute units for one ed25519 signature verification.
-pub const ED25519_VERIFY_COST: u64 = COMPUTE_UNIT_TO_US_RATIO * 76;
+pub const ED25519_VERIFY_COST: u64 = COMPUTE_UNIT_TO_US_RATIO * 80;
```
Comment (tao-stones) [inline cost-model/src/block_cost_limits.rs]: No objection on changing default. It probably should be back ported to 2.0, or put this change behind a feature gate.

**Review by tao-stones — CHANGES_REQUESTED** [review]

File: cost-model/src/block_cost_limits.rs (hunk 1 above)
Comment (samkim-crypto) [reply cost-model/src/block_cost_limits.rs]: Makes sense! I put the changes under the feature gate.

**Review by tao-stones — APPROVED** [review]

**Pull request merged**